### DISCUSSION TITLE
Moving average instead of average/medium for sampling

### DISCUSSION THREAD

**Discussion by sebyon** [OpenGammaProject/Open-Gamma-Detector#34] [Ideas]
Hey, finally gotten some baseline spectra and rough gamma reads, I'm impressed so far. Now my experience is with NIR spectroscopy so I could be totally missing a key fundamental here, if so let me know :) 

I was digging though the software, and noticed at 746

```c++
float avg = 0.0;  // Use median instead of average?
```

I think you're on the right path with that comment, as average alone is sensitive to noise. But as discussed this project is to accommodate many crystal sizes and that will come with increasing or decreasing signal intensities, along with how well the SiPM breakdown voltage and threshold is tuned in. A medium may actually end up causing more noise in some cases. At least with an average it is more forgiving.

Potentially, a approach could be to take a moving average. There are other more exotic methods for signal processing but in the spirit of the project and keeping it computationally friendly.

Excuse me if anything below is incorrect, my C++ is pretty awful. If I were to implement a moving average it would be along the lines of: 

```c++

// Initialize moving average buffer 
const size_t buffer_size = 5; // This will likely need to be balanced between response time and resolution. 
uint16_t buffer[buffer_size];
size_t buffer_index =0;

void eventInt() {
  // ...

  if (!conf.geiger_mode && !adc_lock) {
    uint32_t sum = 0;
    uint8_t num = 0;

    for (size_t i = 0; i < conf.meas_avg; i++) {
      const uint16_t m = analogRead(AIN_PIN);
      // ...

      sum += m;
      num++;

      // Add the new reading to the buffer and update the sum
      sum += m - buffer[buffer_index];
      buffer[buffer_index] = m;
      buffer_index = (buffer_index + 1) % buffer_size;
    }

    float avg = 0.0;

    if (num > 0) {
      avg = float(sum) / float(num);
    }

    // ...
  }

  // ...
}
```
Otherwise I think the current solution of an average as good or a bit better than just the medium. Obviously there are alternatives that would be much better suited, however they're obviously not in line with the concept of this project (external high quality 16+ bit ADC, Savitzky-Golay / Kalman filters...)

**Comment by NuclearPhoenixx** [comment]
Hey, thanks for your ideas! I'm not too sure about all the different implementations and have not come around to test every possibility, that's why the comment is there. So great to have somebody else think about this.

I'm not too sure what the advantages of using a moving average over just the arithmetic mean would be. Is it statistically more robust? I'm not too sure if I understand correctly, especially if we talk about using the same number of ADC measurements for both cases. I'm happy to have a look at a simple moving average and test it if you think it's a better solution.

And yes, you could obviously do all these crazy modifications with 16- or 18-bit ADCs, multiple filters, better software processing and so on... But it's also obvious that with the right setup, the existing electronics can already get pretty good energy resolutions (<10% at least). Generally, multiplying all the processing cost and effort will definitely not return the same amount of performance IMO and will certainly make the device even slower. We're talking about scintillators after all, best you can get is somewhere around 6% as far as I know?

You know this already, but I like to elaborate on this every time something like this comes up, because I get this question quite often. 👍🏻

**Reply by sebyon** [reply]
Honestly, I'm not 100% sure myself and there are a few other simple implementations that might be good too (geometric mean/medium). 

I'm going to have a deep dive in the literature and see how others handle it, I have a feeling they're probably doing something similar to you as it's better practice to leave the data as unprocessed as possible and do the rest on the computer side (like how you do Gaussian correlation filtering with your Gamma-MCA). 

Otherwise the easier test I'll give a try is to disable all averaging, record and import data and then process to see what works best.

**Reply by NuclearPhoenixx** [reply]
Yeah, leaving it as unprocessed as possible is the best way IMO too. The only thing you absolutely need to do at the device level is the ADC measurement(s). Everything else can be done in the MCA software in subsequent processing steps where timing is irrelevant.
